### PR TITLE
Update Container Registry BRS to add support for Ready to Run Remediation Feature

### PR DESCRIPTION
**Pull request by saurabhgautam-msft**

### CODE REVIEW

File: Scripts/RemediationScripts/AutoRemediation.ps1 [hunk 1]
```diff
@@ -153,6 +153,11 @@ Enter the choice (1|2)";
                     $commandString = $control.InitCommand + " -SubscriptionId " +  "`'" + $SubscriptionId +  "`'" +  " -Path " + "`'" + "FailedControls\" +  $SubscriptionId + ".json" + "`'" + " -PerformPreReqCheck"+ " -AutoRemediation" + " -TimeStamp " + "`'" + $timeStampString +  "`'";
                 }elseif ($control.ControlId -eq "Azure_AppService_DP_Use_Secure_TLS_Version") {
                     $commandString = $control.InitCommand + " -SubscriptionId " +  "`'" + $SubscriptionId +  "`'" +  " -Path " + "`'" + "FailedControls\" +  $SubscriptionId + ".json" + "`'" + " -PerformPreReqCheck"+ " -AutoRemediation" + " -TimeStamp " + "`'" + $timeStampString +  "`'";
+                }elseif ($control.ControlId -eq "Azure_ContainerRegistry_Config_Enable_Security_Scanning") {
+                    Write-Host "Object Id for the Security Scanner Identity is required to execute the [$($control.LoadCommand)] Bulk Remediation Script." -ForegroundColor $([Constants]::MessageType.Warning)
```
Comment (asimks) [inline Scripts/RemediationScripts/AutoRemediation.ps1:157]: Give some detail which will help user to find the required object id.

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 2]
```diff
@@ -180,116 +193,192 @@ function Enable-SecurityScanningIdentityForContainerRegistry
...
     )
 
     Write-Host $([Constants]::DoubleDashLine)
-
     if ($PerformPreReqCheck)
     {
         try
         {
-            Write-Host "[Step 1 of 4] Validating and installing the modules required to run the script and validating the user..."
+            Write-Host "[Step 1 of 4] Validating and installing the modules required to run the script and validating the user"
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:216]: Validate and install

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 3]
```diff
@@ -180,116 +193,192 @@ function Enable-SecurityScanningIdentityForContainerRegistry
...
             Setup-Prerequisites
+            Write-Host "Completed setting up prerequisites." -ForegroundColor $([Constants]::MessageType.Update)
+            Write-Host $([Constants]::SingleDashLine)
         }
         catch
         {
-            Write-Host "Error occurred while setting up prerequisites. Error: $($_)" -ForegroundColor $([Constants]::MessageType.Error)
-            break
+            Write-Host "Error occurred while setting up prerequisites. Error: [$($_)]" -ForegroundColor $([Constants]::MessageType.Error)
+            Write-Host $([Constants]::SingleDashLine)
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:227]: Are we using double dash line at end of BRS.

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 4]
```diff
@@ -180,116 +193,192 @@ function Enable-SecurityScanningIdentityForContainerRegistry
...
+            return
         }
     }
     else
     {
-        Write-Host "[Step 1 of 4] Validating the user... "
-    }
-
+        Write-Host "[Step 1 of 4] Validating the user"
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:233]: Validate

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 5]
```diff
@@ -180,116 +193,192 @@ function Enable-SecurityScanningIdentityForContainerRegistry
...
-    else
+    if($AutoRemediation)
     {
-        if (-not (Test-Path -Path $FilePath))
+        if(-not (Test-Path -Path $Path))
         {
-            Write-Host "ERROR: Input file - $($FilePath) not found. Exiting..." -ForegroundColor $([Constants]::MessageType.Error)
-            break
+            Write-Host "Error: File containing failing controls details [$($Path)] not found. Skipping remediation..." -ForegroundColor $([Constants]::MessageType.Error)
+            Write-Host $([Constants]::SingleDashLine)
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:297]: Doble dash at end of execution

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 6]
```diff
@@ -180,116 +193,192 @@ function Enable-SecurityScanningIdentityForContainerRegistry
...
+        Write-Host "Fetching all Container Registry(s) failing for the [$($controlIds)] control from [$($Path)]..." -ForegroundColor $([Constants]::MessageType.Info)
+        Write-Host $([Constants]::SingleDashLine)
+        $controlForRemediation = Get-content -path $Path | ConvertFrom-Json
+        $controls = $controlForRemediation.ControlRemediationList
+        $resourceDetails = $controls | Where-Object { $controlIds -eq $_.ControlId };
+        $validResources = $resourceDetails.FailedResourceList | Where-Object {![String]::IsNullOrWhiteSpace($_.ResourceId)}
+        if(($resourceDetails | Measure-Object).Count -eq 0 -or ($validResources | Measure-Object).Count -eq 0)
+        {
+            Write-Host "No Container Registry(s) found in input json file for remediation." -ForegroundColor $([Constants]::MessageType.Error)
+            Write-Host $([Constants]::SingleDashLine)
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:309]: Doble dash at end of execution

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 7]
```diff
@@ -180,116 +193,192 @@ function Enable-SecurityScanningIdentityForContainerRegistry
...
             {
-                Write-Host "Error fetching Container Registry(s) resource: Resource ID - $($resourceId). Error: $($_)" -ForegroundColor $([Constants]::MessageType.Error)
+                Write-Host "Valid resource id(s) not found in input json file. Error: [$($_)]" -ForegroundColor $([Constants]::MessageType.Error)
+                Write-Host "Skipping the Resource: [$($_.ResourceName)]..." -ForegroundColor $([Constants]::MessageType.Warning)
+                $logResource = @{}
+                $logResource.Add("ResourceGroupName",($_.ResourceGroupName))
+                $logResource.Add("ResourceName",($_.ResourceName))
+                $logResource.Add("Reason","Valid resource id(s) not found in input json file.")    
+                $logSkippedResources += $logResource
+                Write-Host $([Constants]::SingleDashLine)
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:330]: Doble dash line

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 8]
```diff
@@ -299,50 +388,79 @@ function Enable-SecurityScanningIdentityForContainerRegistry
     if ($totalContainerRegistry -eq 0)
     {
         Write-Host "No Container Registry(s) found. Exiting..." -ForegroundColor $([Constants]::MessageType.Warning)
-        break
+        Write-Host $([Constants]::SingleDashLine)
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:391]: Double dash line

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 9]
```diff
@@ -551,31 +762,36 @@ function Disable-SecurityScanningIdentityForContainerRegistry
...
-            break
+            Write-Host "Error occurred while setting up prerequisites. Error: [$($_)]" -ForegroundColor $([Constants]::MessageType.Error)
+            Write-Host $([Constants]::SingleDashLine)
+            return
         }
     }
     else
     {
-        Write-Host "[Step 1 of 3] Validating the user..." 
+        Write-Host "[Step 1 of 3] Validating the user"
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:782]: Validate

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 [hunk 10]
```diff
@@ -700,26 +923,26 @@ function Disable-SecurityScanningIdentityForContainerRegistry
...
             $ContainerRegistrySkipped | Format-Table -Property $colsProperty -Wrap
-            
+            Write-Host $([Constants]::SingleDashLine)
             # Write this to a file.
             $ContainerRegistrySkippedFile = "$($backupFolderPath)\RollbackSkippedContainerRegistry.csv"
             $ContainerRegistrySkipped | Export-CSV -Path $ContainerRegistrySkippedFile -NoTypeInformation
-            Write-Host "This information has been saved to" -NoNewline
-            Write-Host " [$($ContainerRegistrySkippedFile)]" -ForegroundColor $([Constants]::MessageType.Update) 
+            Write-Host "This information has been saved to [$($ContainerRegistrySkippedFile)]" -ForegroundColor $([Constants]::MessageType.Update) 
+            Write-Host $([Constants]::SingleDashLine)
```
Comment (asimks) [inline Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:945]: Double dash line

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 (hunk 7 above)
Comment (saurabhgautam-msft) [reply Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:330]: not needed here as it is not exiting the code at this point

File: Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1 (hunk 10 above)
Comment (saurabhgautam-msft) [reply Scripts/RemediationScripts/Remediate-EnableSecurityScanningForContainerRegistry.ps1:945]: not needed

**Review by asimks — APPROVED** [review]

**Pull request merged**